core/signingrate/dynamo_signing_rate_storage.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	err := s.ensureTableExists(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error ensuring table exists: %w", err)
+	}
+
+	return s, nil
+}
+
+func (d *dynamoSigningRateStorage) StoreBuckets(ctx context.Context, buckets []*validator.SigningRateBucket) error {
+	for _, bucket := range buckets {
+		if err := d.storeBucket(ctx, bucket); err != nil {
+			return fmt.Errorf("error storing bucket: %w", err)
+		}
+	}
+	return nil
+}
+
+func (d *dynamoSigningRateStorage) storeBucket(ctx context.Context, bucket *validator.SigningRateBucket) error {
+
+	key := getDynamoBucketKey(bucket)
+	value, err := proto.Marshal(bucket)
+	if err != nil {
+		return fmt.Errorf("proto marshal failed: %w", err)
+	}
+
+	// Note: the "BucketType" attribute is due a quirk in dynamo. It won't let us do certain queries unless we have
+	// a partition key. So we create a dummy partition key that is always the same value.
+
+	_, err = d.dynamoClient.UpdateItem(ctx, &dynamodb.UpdateItemInput{
+		TableName: d.tableName,
+		Key:       key,
+		UpdateExpression: aws.String(fmt.Sprintf("SET %s = %s, %s = %s, %s = %s",
+			attrPayload, placeholderPayload,
+			attrEndTimestamp, placeholderEndTimestamp,
+			attrPayloadType, placeholderPayloadType)),
+		ExpressionAttributeValues: map[string]types.AttributeValue{
+			placeholderPayload:      &types.AttributeValueMemberB{Value: value},
+			placeholderEndTimestamp: &types.AttributeValueMemberS{Value: timestampToString(bucket.GetEndTimestamp())},
+			placeholderPayloadType:  &types.AttributeValueMemberS{Value: payloadTypeValue},
+		},
+	})
+
+	if err != nil {
+		return fmt.Errorf("dynamo update failed: %w", err)
+	}
+	return nil
+}
+
+// Get the DynamoDB key for a given bucket. The primary key for a bucket is its starting timestamp.
+func getDynamoBucketKey(bucket *validator.SigningRateBucket) map[string]types.AttributeValue {
+	timestamp := bucket.GetStartTimestamp()
+
+	return map[string]types.AttributeValue{
+		attrStartTimestamp: &types.AttributeValueMemberS{Value: timestampToString(timestamp)},
+	}
+}
+
+// Convert a timestamp to the string format used in DynamoDB. String is padded with zeros on the left to ensure
+// lexicographical ordering based on string comparison. This method assumes that timestamps are non-negative and
+// represent seconds since the Unix epoch (i.e. sub-second precision is not supported).
+func timestampToString(unixTime uint64) string {
+	// 20 digits can hold a maximally sized uint64, so ensure that's how much we always use.
+	return fmt.Sprintf("%020d", unixTime)
+}
+
+func (d *dynamoSigningRateStorage) LoadBuckets(
+	ctx context.Context,
+	startTimestamp time.Time,
+) ([]*validator.SigningRateBucket, error) {
+
+	input := &dynamodb.QueryInput{
+		TableName: d.tableName,
+		IndexName: aws.String(endTimestampIndex),
+		KeyConditionExpression: aws.String(fmt.Sprintf("%s = %s AND %s > %s",
+			attrPayloadType, placeholderPayloadType,
+			attrEndTimestamp, placeholderStart)),
```

Comment:
Is this actually correct? Loading all buckets that have an _end_ timestamp that comes after the input _start_ timestamp?

I think for every dynamo query, it would be helpful to have a comment specifying what precisely is being attempted, otherwise it's hard to review correctness of what is actually being done

Reply:
Yes. My intention is to send all data that may have been gathered after a specific timestamp, even if that means also dragging in data that may come from right before that timestamp. Adding more comments.